### PR TITLE
WIP: MDS Spec versus MDS Implementation

### PR DESCRIPTION
**Pull request by ericdill**
- based on conversations with @danielballan, @tacaswell, @arkilic 
  What other changes do we need guys?

### CODE REVIEW

**Comment by ericdill** [comment]
iirc the `event_descriptor` does not have a `type` key anymore, right?

**Comment by ericdill** [comment]
oh, meant to loop @dchabot in on this too, sorry!

**Comment by stuwilkins** [comment]
Hi Eric,

If you look at the top of the document it explains why types are required. While the trigger has gone (I am still a little uncomfortable about that, but @dchabot convinced me) One could think about putting this in the top run header, but I believe this reduces flexibility.

I would suggest not making changes without the small group getting together. @tacaswell made some really good points about how the current metadatastore can be used for different “stages” of data and building in that flexibility at this stage is really important.

Please call a meeting of the few if you would like to change things.

Thanks,
Stuart

On Jan 28, 2015, at 5:38 PM, Eric Dill <notifications@github.com<mailto:notifications@github.com>> wrote:

iirc the event_descriptor does not have a type key anymore, right?

**Comment by ghost** [comment]
It was never part of the descriptor.

On Jan 28, 2015, at 6:03 PM, dchabot <notifications@github.com<mailto:notifications@github.com>> wrote:

File: source/arch/metadatastore-format.rst [hunk 1]
```diff
@@ -171,9 +170,9 @@ above ``event`` is described by the ``event_descriptor`` ::
             "chan6" : {"source" : "PV:XF:23ID1-ES{Sclr:1}.S6"},
             "chan7" : {"source" : "PV:XF:23ID1-ES{Sclr:1}.S7"},
             "chan8" : {"source" : "PV:XF:23ID1-ES{Sclr:1}.S8"}, 
-            "pimte: : {"source" : "FILESTORE:<...>"}
+            "pimte: : {"source" : "name_of_detector", "external": "FILESTORE:<...>"}
```
Comment (tacaswell) [inline source/arch/metadatastore-format.rst]: I thought this is what we decided talking with @stuwilkins

Comment (ghost) [inline source/arch/metadatastore-format.rst]: I think this document is definitely not concrete and always up for discussion. I wouldn't assume anything here is remotely solidified.

File: source/arch/metadatastore-format.rst [hunk 2]
```diff
@@ -161,7 +161,6 @@ above ``event`` is described by the ``event_descriptor`` ::
 
     {
         "uid" : <unique_id>,
-        "type" : "measure",
```
Comment (tacaswell) [inline source/arch/metadatastore-format.rst:164]: I am also not convinced that 'type' makes sense as the name.  The model provides information about what kind of event it is in other ways (what db/collection it is for one thing), type shadows all the things, and does not provide a clear ordering.

I think we should adopt the NASA language here and call this 'data_level' or some such where 0 is raw data.

File: source/arch/metadatastore-format.rst [hunk 3]
```diff
@@ -82,7 +82,7 @@ event. For example::
 
     {
         "uid" : <unique_id>,
-        "scan_id" : <non-unique-id>,
+        "scan_id" : <non-unique-id>, # is this specified to be an integer, or is anything valid here?
```
Comment (tacaswell) [inline source/arch/metadatastore-format.rst]: I would go with anything sortable

Comment (stuwilkins) [inline source/arch/metadatastore-format.rst]: +1

On Jan 29, 2015, at 11:10 AM, Thomas A Caswell <notifications@github.com<mailto:notifications@github.com>> wrote:

In source/arch/metadatastore-format.rsthttps://github.com/NSLS-II/docs/pull/2#discussion_r23778463:

> @@ -82,7 +82,7 @@ event. For example::
> 
> ```
>  {
>      "uid" : <unique_id>,
> ```
> -        "scan_id" : <non-unique-id>,
> -        "scan_id" : <non-unique-id>, # is this specified to be an integer, or is anything valid here?

I would go with anything sortable

**Comment by danielballan** [comment]
I am sitting beside Arman as he finishes implementing this. Here are some decisions we made that deviate from the documented spec:
- Event descriptors are documented to have `keys`, but this clobbers the Python method, so we are reverting to `data_keys`. We don't remember any discussion on this point anyway.
- Sample is currently implemented as a simple dict, not a full-blown collection like Beamline Config. Perhaps in the medium-term it should be an `DynamicEmbeddedDocument`.
- Begin Run Events are indexed on `_id`, `uid`, `time`, `owner`, and `scan_id`. End Run Events are indexed on `_id`, `time`, `exit_status`, and `begin_run_event`.
- Arman prefers the name `exit_status` to `completion_state` and I agree. Can revert later if others dissent.

**Comment by tacaswell** [comment]
These specs shouldn't say anything about indexing

**Comment by tacaswell** [comment]
also, work against #3 not this one

**Pull request merged**